### DISCUSSION TITLE
"npm rebuild" should not run the "preinstall", "install" or "postinstall" scripts

### DISCUSSION THREAD

**Discussion by DeeDeeG** [npm/feedback#218] [CLI] [answered]
### Problem Statement

`npm rebuild` runs the `preinstall`, `install` and `postinstall` scripts as of npm v7.1.1. In my opinion, this makes `npm rebuild` less composable for scripting.

Particularly, this makes my favorite use-case (and intuitively I'd assume this is the main use-case) of `npm rebuild` an infinite loop.

<details>

_(That use-case: In a package's `postinstall` script, right at the point in the lifecycle where the dependencies are all there and ready to rebuild... run `npm rebuild` with another version of Node. This rebuilds the native C++ code against that other version of Node. Unfortunately (in my view), this now calls `postinstall` which calls `npm rebuild` which calls `postinstall`... ♾️)_

</details>

### Potential Solution(s)

- Consider not running any other lifecycle scripts when invoking `npm rebuild`.
- Consider adding some sort of configuration option (I think a flag would be nice) to disable lifecycle scripts for `npm rebuild`. (I tried `npm rebuild --ignore-scripts`, but that skips the rebuilding too!)

<details><summary>More thoughts on this</summary>

I'd like `npm rebuild` to follow a more UNIX philosophy, a "do one thing and do it well" approach. In my opinion, the use-cases for `npm rebuild` are niche, and the context the command is run in are likely to be quite varied. Flexibility is desirable in order to cater to all those use-cases. And in my opinion, proactively running the `preinstall`, `install` and `postinstall` scripts is getting ahead of the user and to reiterate what I mentioned above, diminishes scripting composability.

</details>

### Linking back to the original issue

https://github.com/npm/cli/issues/2670

### New (WIP) lifecycle documentation (for context)

https://github.com/npm/cli/pull/2690

**Comment by MylesBorins** [comment, marked as answer]
@DeeDeeG it looks like you've got a good discussion going on in the npm/cli repo with the team who manage the CLI. There is not much more that we can do over here aside from direct you to work with that team.
